### DISCUSSION TITLE
How to define a separate storage backend?

### DISCUSSION THREAD

**Discussion by joyously** [jj-vcs/jj#2065] [Ideas]
What exactly is involved in defining another backend for commits? (Is that called storage?)

My preferred VCS is Bazaar (latest is called Breezy). I have been able to manage a GitHub repo using Breezy locally, since it knows how to interop with Git.
Is this a matter of using Git as the common denominator, or writing more Rust code to interface with other VCS? Would that be a plugin, or what?

**Comment by martinvonz** [comment]
> Is this a matter of using Git as the common denominator, or writing more Rust code to interface with other VCS? Would that be a plugin, or what?

To define a separate storage backend, you need to write Rust code.  There's an [example here](https://github.com/martinvonz/jj/blob/main/cli/examples/custom-backend/main.rs).

However, if you're asking if you can use jj with your existing git repo (which you also use Breezy with), then you don't need a new storage backend, because we already have a storage backend for it.

Does that answer your question?

**Comment by joyously** [comment]
It answers it somewhat.
The example is helpful, but I would still not know what to change to use Breezy or Darcs.

I could use jj with that one repo I was using with GitHub, but most of my repos are not associated with Git.

**Reply by martinvonz** [reply]
> The example is helpful, but I would still not know what to change to use Breezy or Darcs.

Darcs's model is very different, so it's probably going to be very hard to fit it into jj's snapshot-based and git-like commits/trees/files model. I don't know enough about Breezy to be sure, but given that it has a git backend, it must be somewhat similar. I wouldn't be surprised if it's still hard to get it work, however. Mercurial's model is also quite similar to git's, and making a read-only backend for it is easy. The flow for creating commits in Mercurial is quite different, however, which makes it hard create a writable implementation. On problem is that each file changed in a commit is expected to have pointer back to a commit, which doesn't mesh with creating the commit afterwards (or not at all!) like jj and git do.

**Comment by joyously** [comment]
Yes, I'm aware that Darcs is different, but it seems that you could just feed it the diff.
Breezy is snapshot-based and already works with Git (it does not have a Git backend).
Since Breezy is Python, would everything be run through external command? If so, it seems like that could be made generic and anyone could hook up a backend with a config file of commands to use.

**Comment by martinvonz** [comment]
The example I linked to earlier shows which functions need to be implemented by a backend. It's up to the backend how it provides that functionality. The git backend read and writes to a backing git repo. The native backend stores its data in protobuf files. The backend we have at Google uses RPCs. A backend certainly could run a Python command for every object it's asked to read or write. That should work, but it would probably be very slow for anything but a very small repo.

**Comment by joyously** [comment]
I thought I should update this to say that in an experiment using a Git repo, I was able to use most commands from Git, Breezy, or Jujutsu and it was reflected in the log in all three. This was after `git init` and `jj git init` (Breezy didn't need `init`). 
However, when i used `git clone` to a new folder, I would have to init `jj` again there. I understand why, but it was a jolt when I could use Breezy with no preparation in any Git repo.
